File: src/Asm.MooBank.Domain/Entities/Account/RecurringTransaction.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Diagnostics.CodeAnalysis;
using Asm.Domain;
using Microsoft.EntityFrameworkCore;

namespace Asm.MooBank.Domain.Entities.Account;

public class RecurringTransaction([DisallowNull] Guid id) : KeyedEntity<Guid>(id)
{
    public RecurringTransaction() : this(default) { }
    public Guid VirtualAccountId { get; set; }
    public string? Description { get; set; }

    [Precision(12, 4)]
    public decimal Amount { get; set; }
    public DateTime? LastRun { get; set; }

    public DateOnly NextRun { get; set; } = DateTime.UtcNow.ToDateOnly();

    public virtual VirtualInstrument VirtualAccount { get; set; } = null!;

    [Column("ScheduleId")]
    public virtual ScheduleFrequency Schedule { get; set; }
}

File: src/Asm.MooBank.Web.App/src/models/RecurringTransaction.ts
import { emptyGuid } from "@andrewmclachlan/mooapp";

export interface RecurringTransaction {
    id: string;
    description?: string;
    amount: number;
    schedule: Schedule;
    lastRun?: string;
    nextRun: string;
    virtualAccountId: string;
}

export const emptyRecurringTransaction = (virtualAccountId: string): RecurringTransaction => ( {
    id: emptyGuid,
    description: "",
    amount: 0,
    schedule: "Weekly",
    lastRun: null,
    nextRun: "",
    virtualAccountId: virtualAccountId,
});

export const Schedules = ["Daily", "Weekly", "Monthly",  "Yearly"] as Schedule[];
export type Schedule = "Daily" | "Weekly" | "Monthly" | "Yearly";